lib/handler/pool.ex

```diff
@@ -1,58 +1,142 @@
 defmodule Handler.Pool do
-  defstruct max_workers: 100,
+  defstruct delegate_to: nil,
+            max_workers: 100,
             max_memory_bytes: 10 * 1024 * 1024 * 1024,
             name: nil
 
   alias __MODULE__
   alias Handler.Pool.State
   use GenServer
 
-  def start_link(%Pool{} = config) do
-    GenServer.start_link(Pool, config)
+  @type t :: %Handler.Pool{
+          delegate_to: nil | name(),
+          max_workers: non_neg_integer(),
+          max_memory_bytes: non_neg_integer(),
+          name: nil | name()
+        }
+  @type name :: GenServer.name()
+  @type pool :: GenServer.server()
+  @type exception :: Pool.InsufficientMemory.t() | Pool.NoWorkersAvailable.t()
+
+  @moduledoc """
+  Manage a pool of resources used to run dangeours functions
+
+  The Pool provides both a synchrous interface (`run/3`) as well as an asynchronous
+  interface (`async/3` and `await/1`) for running functions.
+
+  ## Composing Pools
+
+  This module also provides a way to have limited resources for particular use-cases, that end up
+  sharing bigger pools of resources. Take for instance a hosted multi-tenant application where you
+  can safely use up to 10GB of memory for a particular task, but you don't want any one customer to
+  use the whole pool, so each customer has a limit of just 5GB. You can handle this scenario using
+  the `delegate_to` configuration.
+
+      {:ok, _pid} = Handler.Pool.start_link(%Handler.Pool{
+        max_memory: 10 * 1024 * 1024 * 1024,
```

Comment:
Maybe we should add some default units as constants (e.g. make a `Handler.gigabyte` function). Just a thought.